core/includes/install.core.inc

```diff
@@ -1454,6 +1450,340 @@ function install_select_language_form($form, &$form_state, $files) {
+
+  $data = '';
+  if ($response->code === '200') {
+    $data = json_decode($response->data, TRUE);
+  }
+
+  if ($data) {
+    $available_translations = array_keys($data);
+  }
+  elseif (count($files) > 1) {
+    foreach($files as $translation_file) {
+      $available_translations[] = $translation_file->langcode;
+    }
+  }
+  else {
+    return FALSE;
+  }
+
+  // English is always available.
+  $available_translations[] = 'en';
+  $standard_languages = standard_language_list();
+  return array_intersect_key($standard_languages, array_flip(array_unique($available_translations)));
+}
+
+/**
+ * Download a translation file for the selected language.
+ *
+ * @param array $install_state
+ *   An array of information about the current installation state.
+ *
+ * @return string
+ *   A themed status report, or an exception if there are requirement errors.
+ *   Upon successful download the page is reloaded and no output is returned.
+ */
+function install_download_translation(&$install_state) {
+  // Check whether all conditions are met to download. Download the translation
+  // if possible.
+  $requirements = install_check_translations($install_state['parameters']['langcode'], $install_state['server_pattern']);
+  // Render requirements if any warnings or errors returned.
+  if ($output = install_display_requirements($install_state, $requirements)) {
+    return $output;
+  }
+
+  // The download was successful, reload the page in the new language.
+  $install_state['translations'][$install_state['parameters']['langcode']] = TRUE;
+}
+
+/**
+ * Attempts to get a file using a HTTP request and to store it locally.
+ *
+ * @param string $uri
+ *   The URI of the file to grab.
+ * @param string $destination
+ *   Stream wrapper URI specifying where the file should be placed. If a
+ *   directory path is provided, the file is saved into that directory under its
+ *   original name. If the path contains a filename as well, that one will be
+ *   used instead.
+ *
+ * @return bool
+ *   TRUE on success, FALSE on failure.
+ */
+function install_retrieve_file($uri, $destination) {
+  $parsed_url = parse_url($uri);
+
+  if (is_dir(backdrop_realpath($destination))) {
+    // Prevent URIs with triple slashes when gluing parts together.
+    $path = str_replace('///', '//', "$destination/") . backdrop_basename($parsed_url['path']);
+  }
+  else {
+    $path = $destination;
+  }
+
+  $response = backdrop_http_request($uri);
+  if ($response->code != 200) {
+    return FALSE;
+  }
+
+  $data = $response->data;
+  if (empty($data)) {
+    return FALSE;
+  }
+
+  return file_put_contents($path, $data) !== FALSE;
+}
+
+/**
+ * Checks if the localization server can be contacted.
+ *
+ * @param string $uri
+ *   The URI to contact.
+ *
+ * @return string
+ *   TRUE if the URI was contacted successfully, FALSE if not.
+ */
+function install_check_localization_server($uri) {
+  $response = backdrop_http_request($uri);
```

Comment:
As we're only checking for a bool here, it would also be possible to just use a lightweight HEAD request - without transferring the json data.

```
$options = array('method' => 'HEAD');
$response = backdrop_http_request($uri, $options);
```
A GET request also does the job, of course.

Reply:
Done